debugging/coreclr/dockerClient.ts

```diff
@@ -0,0 +1,194 @@
+
+export type DockerBuildImageOptions = {
+    context?: string;
+    dockerfile?: string;
+    tag?: string;
+    target?: string;
+};
+
+export type DockerInspectObjectOptions = {
+    format?: string;
+};
+
+export type DockerContainersListOptions = {
+    format?: string;
+};
+
+export type DockerContainerRemoveOptions = {
+    force?: boolean;
+};
+
+export type DockerContainerVolume = {
+    localPath: string;
+    containerPath: string;
+    permissions?: 'ro' | 'rw';
+};
+
+export type DockerRunContainerOptions = {
+    command?: string;
+    containerName?: string;
+    entrypoint?: string;
+    volumes?: DockerContainerVolume[];
+};
+
+export interface DockerClient {
+    buildImage(options: DockerBuildImageOptions, progress?: (content: string) => void): Promise<string>;
+    inspectObject(nameOrId: string, options?: DockerInspectObjectOptions): Promise<string | undefined>;
+    listContainers(options?: DockerContainersListOptions): Promise<string>;
+    matchId(id1: string, id2: string): boolean;
+    removeContainer(containerNameOrId: string, options?: DockerContainerRemoveOptions): Promise<void>;
+    runContainer(imageTagOrId: string, options?: DockerRunContainerOptions): Promise<string>;
+    trimId(id: string): string;
+}
+
+export class CliDockerClient implements DockerClient {
+    constructor(private readonly processProvider: ProcessProvider) {
+    }
+
+    public async buildImage(options?: DockerBuildImageOptions, progress?: (content: string) => void): Promise<string> {
+        let command = `docker build --rm`;
+
+        if (options && options.dockerfile) {
+            command += ` -f ${options.dockerfile}`;
+        }
+
+        if (options && options.tag) {
+            command += ` -t ${options.tag}`;
+        }
+
+        if (options && options.target) {
+            command += ` --target ${options.target}`;
+        }
+
+        if (options && options.context) {
+            command += ` ${options.context}`;
+        }
+
+        let imageId: string | undefined;
+
+        // TODO: Handle case where content may not be whole lines.
+        const buildProgress =
+            (content: string) => {
+                if (progress) {
+                    progress(content);
+                }
+
+                // Last line of expected output is: 'Successfully built 7cc5654ca3b6'
+                const buildSuccessPrefix = 'Successfully built ';
+
+                if (content.startsWith(buildSuccessPrefix)) {
+                    imageId = content.substr(buildSuccessPrefix.length, 12);
+                }
+            };
+
+        await this.processProvider.exec(command, { progress: buildProgress });
+
+        if (!imageId) {
+            throw new Error('The Docker image was built successfully but the image ID could not be retrieved.');
+        }
+
+        return imageId;
+    }
+
+    public async inspectObject(nameOrId: string, options?: DockerInspectObjectOptions): Promise<string | undefined> {
```

Comment:
Noted and I think that can *probably* be accommodated.  For VS we don't explicitly support debugging via non-local Docker hosts, though it will likely work assuming you've setup the Docker CLI context to default to the remote host.  Since debugging requires passing the Core CLR debugging an actual Docker command line, we'd need to ensure that the Docker host used by `docker-endpoint` matches up with that command line used for debugging.

In any case, I would prefer to make such a swap as a separate PR to simplify testing.  I've added a `CONSIDER` comment to `dockerClient`.